lib/screens/v2/candidate_generator/dup/departures.ex

```diff
@@ -501,37 +501,76 @@ defmodule Screens.V2.CandidateGenerator.Dup.Departures do
 
   defp get_overnight_schedules_for_section(_, _, _, _, _, _, _), do: []
 
+  defp get_overnight_departure_for_route(
+         _first_schedule_today,
+         nil,
+         _first_schedule_tomorrow,
+         route_id,
+         direction_id,
+         now
+       ) do
+    Logger.warn(
+      "[get_overnight_schedules_for_section] last_schedule_today not found. route_id=#{route_id} direction_id=#{direction_id} now=#{now}"
+    )
+
+    nil
+  end
+
+  defp get_overnight_departure_for_route(
+         first_schedule_today,
+         last_schedule_today,
+         first_schedule_tomorrow,
+         route_id,
+         direction_id,
+         now
+       ) do
+    cond do
+      is_nil(first_schedule_tomorrow) ->
+        nil
+
+      DateTime.compare(now, first_schedule_tomorrow.departure_time) == :gt ->
+        Logger.warn(
+          "[get_overnight_schedules_for_section] now is after first_schedule_tomorrow. route_id=#{route_id} direction_id=#{direction_id} now=#{now}"
+        )
+
+        nil
+
+      # Before 3am and between the `departure_time` for today's last schedule and tomorrow's first schedule
+      DateTime.compare(now, last_schedule_today.departure_time) == :gt and
+          DateTime.compare(now, first_schedule_tomorrow.departure_time) == :lt ->
+        %Departure{schedule: first_schedule_tomorrow}
+
+      # After 3am but before the first scheduled trip of the day.
```

Comment:
For these two cases, how are we certain that it's before 3am if we're only checking `now` against the `last_schedule_today` and the `first_schedule_tomorrow`?

Reply:
Hmm okay I think I'm following. Essentially we're inferring if it's currently before or after 3am based on how `now` relates to `last_schedule_today` and `first_schedule_tomorrow`? Or I guess mainly just `last_schedule_today` to determine that, and then we compare against `first_schedule_tomorrow` to know if we should show overnight mode?

Reply:
Yep, that's right!